### PR TITLE
CRM-21461: Case Dashlet enhancement

### PR DESCRIPTION
**Pull request by monishdeb**
Overview
----------------------------------------
1. Convert Case dashlets 'All Cases'/'My Cases' with AJAX selectors to use DataTable for listing
2. Add Case filters for search cases by status, type and 'Search Cases with Upcoming Activities' checkbox (unchecking will show cases with 'any' activities)
3. Supported sorting by contact, subject, date, status, type and role.
4. Editable Case subject.

Before
----------------------------------------
Case Dashboard:
screen shot 2017-11-28 at 11 20 31 am

Case Dashlets:
screen shot 2017-11-28 at 11 20 16 am

After
----------------------------------------
Case Dashboard:
screen shot 2017-11-28 at 11 18 20 am

Case Dashlets:
screen shot 2017-11-28 at 11 17 30 am

---

 * [CRM-21461: Case Dashlet enhancement](https://issues.civicrm.org/jira/browse/CRM-21461)

### CODE REVIEW

File: templates/CRM/Dashlet/Page/AllCases.tpl [hunk 1]
```diff
@@ -23,10 +23,11 @@
  | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
  +--------------------------------------------------------------------+
 *}
-{if $AllCases}
-   <div class="form-item">
-       {include file="CRM/Case/Page/DashboardSelector.tpl" context="$context" list="allcases" rows=$AllCases}
-   </div>
+{if $casePresent}
```
Comment (lcdservices) [inline templates/CRM/Dashlet/Page/AllCases.tpl]: the logic is flawed here (and in MyCases.tpl). it only shows the filter if cases are present. however, that means you're locked into the default filters with no way to change it. so if, for example, there are no cases with upcoming activities in the next two weeks, you have no way to change the filter to show all cases. I think we can probably just remove the "no open cases"  clause altogether and persistently show the filter block.

File: CRM/Case/BAO/Case.php [hunk 2]
```diff
@@ -530,17 +537,27 @@ public static function getCaseActivityQuery($type = 'upcoming', $userID = NULL,
 
     if ($condition) {
       // CRM-8749 backwards compatibility - callers of this function expect to start $condition with "AND"
-      $query .= " WHERE (1) $condition ";
+      $query .= " WHERE (1) AND $condition ";
```
Comment (lcdservices) [inline CRM/Case/BAO/Case.php]: the AND threw an error. see getNextScheduledActivity -- the $condition that's passed to this already has an AND prefixed, resulting in a double AND. actually -- that error occurred in the contact's case tab. removing the AND throws errors in the dashlet. so we need to do some code review to make sure $condition is passed to this function in a consistent way.

File: templates/CRM/Dashlet/Page/AllCases.tpl (hunk 1 above)
Comment (monishdeb) [reply templates/CRM/Dashlet/Page/AllCases.tpl]: ok will remove this condtion.

File: CRM/Case/BAO/Case.php (hunk 2 above)
Comment (monishdeb) [reply CRM/Case/BAO/Case.php]: Ahh ok, my bad will fix it and will check its usage on other places to avoid regression.

**Comment by monishdeb** [comment]
Jenkins test this please

File: templates/CRM/Case/Page/DashboardSelector.tpl [hunk 3]
```diff
@@ -25,74 +25,49 @@
...
+    <th data-data="activity_list" data-orderable="false" class="crm-case-activity_list"></th>
+    <th data-data="contact_id" class="crm-case-contact">{ts}Contact{/ts}</th>
+    <th data-data="subject" cell-class="crmf-subject crm-editable" class="crm-case-subject">{ts}Subject{/ts}</th>
+    <th data-data="case_status" class="crm-case-status">{ts}Status{/ts}</th>
+    <th data-data="case_type" class="crm-case-type">{ts}Type{/ts}</th>
+    <th data-data="case_role" class="crm-case-role">{ts}My Role{/ts}</th>
+    <th data-data="manager" data-orderable="false" class="crm-case-manager">{ts}Manager{/ts}</th>
+    <th data-data="date" cell-class="crm-case-date">{if $list EQ 'upcoming'}{ts}Next Sched.{/ts}{elseif $list EQ 'recent'}{ts}Most Recent{/ts}{/if}</th>
+    <th data-data="links" data-orderable="false" class="crm-case-links">&nbsp;</th>
+  </th>
```
Comment (lcdservices) [inline templates/CRM/Case/Page/DashboardSelector.tpl]: closing tag should be </tr>

**Comment by lcdservices** [comment]
this looks great. dashlets are much more dynamic and flexible now. @colemanw can you review and merge.

**Comment by monishdeb** [comment]
@colemanw / @mlutfy can you please review this PR?

**Comment by colemanw** [comment]
Code looks great. Merging based on Brian's testing.

**Pull request merged**